java/src/main/java/ai/rapids/cudf/MemoryCleaner.java

```diff
@@ -214,10 +214,24 @@ static void setDefaultGpu(int defaultGpuId) {
         if (defaultGpu >= 0) {
           Cuda.setDevice(defaultGpu);
         }
+
         for (CleanerWeakReference cwr : all) {
           cwr.clean();
         }
-      }));
+      };
+
+      // Shutdown hooks are executed concurrently in JVM, and there is no execution order guarantee.
+      // See the doc of `Runtime.addShutdownHook`.
+      // Some resources are closed in other Spark hooks.
+      // Here we should wait other Spark hooks to be done, or a false leak will be detected.
+      //
+      // `Spark ShutdownHookManager` leverages `Hadoop ShutdownHookManager` to manage hooks with
+      // priority. The priority parameter will guarantee the execution order.
+      //
+      // Here also use `Hadoop ShutdownHookManager` to add a lower priority hook.
+      // 20 priority is small enough, will run after Spark hooks.
+      // Note: `ShutdownHookManager.get()` is a singleton, Spark and JNI both use the same singleton
+      org.apache.hadoop.util.ShutdownHookManager.get().addShutdownHook(hook, 20);
```

Comment:
I really don't like using the Hadoop ShutdownHookManager here. To solve the problem all of the shutdown hooks that might conflict with each other must use this manager too.  Which in some cases they do, but in other cases they will not. This is a library and it is not Hadoop specific.

I don't want to over engineer this, but I personally would like to see a way for the user of the library to provide the way for the shutdown code to run, and if none is provided then the regular java Runtime is used. But this is hard because all of the code is running inside a static block. We might need to first register the hook with the java Runtime and then remove it if/when someone tells us to use a different shutdown mechanism.

Reply:
1/ We can create a Runnable hook without automatically adding it to `Runtime`.
 
2/ We can add a Boolean system property for whether to add shutdown hook directly to the `Runtime`. 

3/ Then have a contract MemroryCleaner.getCleanerHook() that returns non-null from 1 if the Boolean from 2 is false

Reply:
Already had one Boolean `REF_COUNT_DEBUG`: 
https://github.com/rapidsai/cudf/blob/branch-22.08/java/src/main/java/ai/rapids/cudf/MemoryCleaner.java#L203
```
    if (REF_COUNT_DEBUG) {
      // If we are debugging things do a best effort to check for leaks at the end
      Runtime.getRuntime().addShutdownHook(new Thread(() -> {
......
        for (CleanerWeakReference cwr : all) {
          cwr.clean();
        }
      }));
    }
```

If disable `REF_COUNT_DEBUG`,  there are no fake error logs.
We can create a new one like `REF_COUNT_DEBUG_WHEN_CLOSE` to replace it.
But still have this problem if enables the boolean.

Reply:
In my interpretation `ai.rapids.refcount.debug` is here to determine whether to check for leaks at shutdown time, and not about how to achieve this.

So either another Boolean switch to the tune of `ai.rapids.refcount.debug.addShutdownHook` 
or we can change `ai.rapids.refcount.debug` to a String property with valid values to the tune of: "none", "withInternalShutdownHook", "withExternalShutdownHook" .  We can use an Enum to [manage them](https://stackoverflow.com/a/604426).